Sources/Datadog/Core/System/CarrierInfoProvider.swift

```diff
@@ -66,88 +66,145 @@ extension CarrierInfo.RadioAccessTechnology {
-        publisher.subscribe(subscriber)
+        wrappedProvider.subscribe(subscriber)
     }
 }
 
 #if targetEnvironment(macCatalyst)
 
-internal struct MacCatalystCarrierInfoProvider: WrappedCarrierInfoProvider {
-    /// Carrier info is not supported on macCatalyst
+/// Dummy provider for Mac Catalyst which doesn't support carrier info.
+internal struct MacCatalystCarrierInfoProvider: CarrierInfoProviderType {
     var current: CarrierInfo? { return nil }
+    func subscribe<Observer>(_ subscriber: Observer) where Observer: ValueObserver, Observer.ObservedValue == CarrierInfo? {}
 }
 
 #else
 
-internal struct iOSCarrierInfoProvider: WrappedCarrierInfoProvider {
-    let networkInfo: CTTelephonyNetworkInfo
+/// Carrier info provider for iOS 12 and above.
+/// It reads `CarrierInfo?` from `CTTelephonyNetworkInfo` only when `CTCarrier` has changed (e.g. when the SIM card was swapped).
+@available(iOS 12, *)
+internal class iOS12CarrierInfoProvider: CarrierInfoProviderType {
+    private let networkInfo: CTTelephonyNetworkInfo
+    /// Publisher for notifying observers on `CarrierInfo` change.
+    private let publisher: ValuePublisher<CarrierInfo?>
 
-    var current: CarrierInfo? {
-        let carrier: CTCarrier?
-        let radioTechnology: String?
-
-        if #available(iOS 12, *) {
-            guard let cellularProviderKey = networkInfo.serviceCurrentRadioAccessTechnology?.keys.first else {
-                return nil
-            }
-            radioTechnology = networkInfo.serviceCurrentRadioAccessTechnology?[cellularProviderKey]
-            carrier = networkInfo.serviceSubscriberCellularProviders?[cellularProviderKey]
-        } else {
-            radioTechnology = networkInfo.currentRadioAccessTechnology
-            carrier = networkInfo.subscriberCellularProvider
+    init(networkInfo: CTTelephonyNetworkInfo) {
+        self.networkInfo = networkInfo
+        self.publisher = ValuePublisher(
+            initialValue: iOS12CarrierInfoProvider.readCarrierInfo(
+                from: networkInfo,
+                cellularProviderKey: networkInfo.serviceCurrentRadioAccessTechnology?.keys.first
+            )
+        )
+
+        // The `serviceSubscriberCellularProvidersDidUpdateNotifier` block object executes on the default priority
+        // global dispatch queue when the user’s cellular provider information changes.
+        // This occurs, for example, if a user swaps the device’s SIM card with one from another provider, while the app is running.
+        // ref.: https://developer.apple.com/documentation/coretelephony/cttelephonynetworkinfo/3024512-servicesubscribercellularprovide
+        networkInfo.serviceSubscriberCellularProvidersDidUpdateNotifier = { [weak self] cellularProviderKey in
+            let carrierInfo = iOS12CarrierInfoProvider.readCarrierInfo(
+                from: networkInfo,
+                cellularProviderKey: networkInfo.serviceCurrentRadioAccessTechnology?.keys.first
+            )
```

Comment:
I've been debugging what happens to `networkInfo: CTTelephonyNetworkInfo` when removing and changing SIM card. The object itself is not deallocated, only its values are updated:
<img width="1001" alt="Screenshot 2021-10-06 at 18 43 46" src="https://user-images.githubusercontent.com/2358722/136374048-5a493182-6e72-43f7-abb0-0758cdb84233.png">

What changes though is the `CTCarrier` retrieved in `readCarrierInfo(from:cellularProviderKey:)`:
```swift
let carrier: CTCarrier = networkInfo.serviceSubscriberCellularProviders?[cellularProviderKey]
```
<img width="1409" alt="Screenshot 2021-10-07 at 13 40 44" src="https://user-images.githubusercontent.com/2358722/136377408-b2ff34bb-4c77-49a9-8db2-839cde2e4e0c.png">


The `swift_unknownObjectRetain` in reported crash:
```
0  libswiftCore.dylib             0x18ed94b70 swift_unknownObjectRetain + 20
1  libswiftFoundation.dylib       0x18e8b7f48 closure #2 in static Dictionary._forceBridgeFromObjectiveC(_:result:) + 736
2  libswiftFoundation.dylib       0x18e8b8000 partial apply for closure #2 in static Dictionary._forceBridgeFromObjectiveC(_:result:) + 36
3  libswiftFoundation.dylib       0x18e8bd32c _NativeDictionary.init(_unsafeUninitializedCapacity:allowingDuplicates:initializingWith:) + 340
4  libswiftFoundation.dylib       0x18e8b6ad0 static Dictionary._unconditionallyBridgeFromObjectiveC(_:) + 1368
5  Datadog                        0x103c948b8 iOSCarrierInfoProvider.current.getter + 132 (CarrierInfoProvider.swift:132)
```
seems to correspond to extending the `CTCarrier's` lifespan in our code after it was released in `CoreTelephony`. This is one of the scenarios I've been trying to reproduce (extending it explicitly), but it didn't lead to any crash.

Reply:
This creates a retain cycle of `networkInfo`.

Reply:
Why not using the `cellularProviderKey` from the `serviceSubscriberCellularProvidersDidUpdateNotifier` callback?

Reply:
> I see a retain cycle of networkInfo here.

Oh yeah, my bad. It doesn't make any bad side effects, as `networkInfo` is a singleton (once obtained from `CoreTelephony` it never changes). Anyway, will fix that, good catch 👍 .

> Also, why not using the `cellularProviderKey` returned by the `serviceSubscriberCellularProvidersDidUpdateNotifier` closure?

My bad, again, good catch 👍. In all my tests, it has the same value as `serviceCurrentRadioAccessTechnology's` `.keys.first`, but the entire point for having this callback is to indeed use `cellularProviderKey` instead. Will fix.

Reply:
Fixed 👍. I did one more test, it works.

<img width="511" alt="Screenshot 2021-10-08 at 16 05 05" src="https://user-images.githubusercontent.com/2358722/136571029-e373c1f7-1b2e-4b42-9ba1-d877a41a9751.png">